### DISCUSSION TITLE
Which RL algorithm is most optimal for SAP?

### DISCUSSION THREAD

**Discussion by andreped** [andreped/super-ml-pets#10] [Ideas]
From the discussion here https://github.com/GoldExplosion/SuperAutoPets-RL-Agent/issues/1#issuecomment-1218553044

It was clearly that the choice of RL design for this problem is not trivial. This thread serves as a source for discussing designs and brainstorming idea.

**Comment by GoldExplosion** [comment]
Coming up with a non RL design as a baseline helps with comparing different algorithms and also serves as an example on why RL is required for this problem. Coming up with a non RL design for a imperfect information game is hard but other games like poker (an imperfect information game) have come up with such design which we can draw inspiration from.

**Reply by andreped** [reply]
I like that idea. At least to serve as a sanity check to assess that the model is doing what it should doing during training.

**Comment by GoldExplosion** [comment]
sapai gym during training gives reward based on how well the algorithm does against a randomly generated opponent. We can make a better opponent by playing the game ourselves and noting down the team we create at the end of each turns. Then we create an opponent that will serve those team we recorded. This increases the difficulty of the opponent and thus pushes the algorithm to do better for more reward.

**Reply by andreped** [reply]
Why doing we do a Q-learning based approach, where the AI somewhat play against itself? Randomly generated is somewhat naive. We should also give the "opponent" the option to build a team, similar to what the AI is doing.

I was initially thinking of storing real game, but this takes **a lot** of games to be a viable option. I would rather do a Q-learning based approach, as it requires very little to do. By doing Q-learning the algorithm naturally gets better by naturally meeting better opponents (previous states).

**Comment by GoldExplosion** [comment]
Q-Learning is a great algorithm for problems which involve a small state space (like tik tac toe) but not practical to use on problems with a huge state space (like chess) because the q learning table will be unreasonably large. To solve big state space problems, Deep Q-Network was introduced. In DQN, the q value is not calculated but estimated. 

Super Auto Pets has a large state space. In sapai, manny405 has done something similar where he ranked teams based on win rate for the 1st round and there were 5013 states or teams (https://github.com/manny405/sapai/tree/main/data) he had to compare. This is the number of states for 8 pets and 2 food. By the time you reach the final tier, you have 59 pets and  17 food. The corresponding state will be in the millions I suppose. 

We can try storing 10 real games and have the opponent choose randomly from these games to emulate. I don't see why it requires a lot of games for it to be viable.

**Reply by andreped** [reply]
I didnt mean that we swap the MaskablePPO with Q-Learning (and of course, I also meant DQN), but rather that we use a Q-Learning-inspired opponent generation.

But I mean, there are lots of ways to approach this problem. Lets just explore ideas and test them, and report what works and what doesn't. No need to choose only one tactic :)

And I like your idea, I just found it more annoying to setup :P But of course, the current training scheme is a bit naive. We should do something about it.

**Reply by GoldExplosion** [reply]
I agree with you. Sorry if i seemed to push too hard. 

There is a way to reduce the number of states in the Q-Learning approach for opponent generation. Basically, when the team is reordered it is considered a new state. We can compare different arrangements of Pets in the team and only consider the top arrangements. This will significantly reduce the number of states by a factor of (no of pets in the team)!

**Reply by andreped** [reply]
> I agree with you. Sorry if i seemed to push too hard.

Hehe, no worries :P



> only consider the top arrangements

From a curriculum-learning aspect this makes a lot of sense. Perhaps at the beginning we can allow random arrangement of opponents, but then we can introduce more challenging opponents. Lots of ways to approach this.

As I am currently mostly trying to get the current solution stable and working across python versions and platforms, I don't have time to look into this, but perhaps you can? :) Should work in a separate branch and make a PR when you have something working, and I could test it :)

**Reply by GoldExplosion** [reply]
yes, I will work on this.

**Reply by andreped** [reply]
@GoldExplosion Did you get the deployment stuff working? I made lots of bug fixes, and at least now if I deploy (and finetune) a model trained on python 3.7, I can use it on a separate machine with python 3.8. It should handle different python versions as well, as long as the python version you use for deployment has the same or newer version than the model was trained with.

I added lots of comments to this here: https://github.com/andreped/super-ml-pets/issues/16

**Reply by GoldExplosion** [reply]
Thanks I just got it working.
